src/transformers/api-history.ts

```diff
@@ -0,0 +1,318 @@
+  endCursor: string;
+}
+
+type GithubArtifactsListResponse = {
+  artifacts: Array<{ id: number; name: string; archive_download_url: string }>;
+};
+
+function isObject(
+  value: unknown
+): value is Record<string | number | symbol, unknown> {
+  return typeof value === 'object' && value !== null && !Array.isArray(value);
+}
+
+function isGithubArtifactsListResponse(
+  value: unknown
+): value is GithubArtifactsListResponse {
+  return isObject(value) && 'artifacts' in value;
+}
+
+function isPrReleaseArtifact(value: unknown): value is PrReleaseArtifact {
+  return isObject(value) && 'data' in value && 'endCursor' in value;
+}
+
+function isApiHistory(value: unknown): value is ApiHistory {
+  return (
+    isObject(value) &&
+    ('added' in value || 'deprecated' in value || 'changes' in value)
+  );
+}
+
+export default function attacher() {
+  return transformer;
+}
+
+function matchApiHistoryCodeBlock(node: Node): node is Code {
+  return (
+    isCode(node) &&
+    node.lang?.toLowerCase() === 'yaml' &&
+    node.meta?.toLowerCase() === 'history'
+  );
+}
+
+let _allElectronVersions: SemVer[];
+
+async function getAllElectronVersions(): Promise<SemVer[]> {
+  if (_allElectronVersions) {
+    return _allElectronVersions;
+  }
+
+  const { versions } = await ElectronVersions.create(undefined, {
+    ignoreCache: true,
+  });
+
+  if (versions == null) throw new Error('No Electron versions found.');
+
+  _allElectronVersions = versions;
+  return _allElectronVersions;
+}
+
+let _allPrReleaseVersions: PrReleaseVersionsContainer;
+
+async function getAllPrReleaseVersions(): Promise<PrReleaseVersionsContainer> {
+  if (_allPrReleaseVersions) {
+    return _allPrReleaseVersions;
+  }
+
+  // TODO: Remove this
+  if (!process.env.GH_TOKEN) {
+    const versions: PrReleaseVersionsContainer = {
+      35658: {
+        release: 'v30.0.0-nightly.20231214',
+        backports: ['v29.0.0-alpha.9'],
+      },
+      41391: {
+        release: 'v31.0.0-alpha.1',
+        backports: [] as string[],
+      },
+      42086: {
+        release: 'v32.0.0-nightly.20240531',
+        backports: ['v31.0.0'],
+      },
+    };
+
+    _allPrReleaseVersions = versions;
+    return _allPrReleaseVersions;
+  }
+
+  if (!process.env.GH_TOKEN) {
+    logger.warn(
+      'No GitHub token found, skipping fetching PR release versions.'
+    );
+    _allPrReleaseVersions = {};
+    return _allPrReleaseVersions;
+  }
+
+  const fetchOptions = {
+    method: 'GET',
+    headers: {
+      Accept: 'application/vnd.github+json',
+      'X-GitHub-Api-Version': '2022-11-28',
+      Authorization: `Bearer ${process.env.GH_TOKEN}`,
+    },
+  };
+
+  const artifactsListResponse = await fetch(
+    GH_ACTIONS_ARTIFACTS_URL,
+    fetchOptions
+  );
+  const artifactsListResponseJson = await artifactsListResponse.json();
+
+  if (!isGithubArtifactsListResponse(artifactsListResponseJson))
+    throw new Error('Invalid GitHub artifacts list response.');
```

Comment:
```suggestion
  if (!isGithubArtifactsListResponse(artifactsListResponseJson)) {
    throw new Error('Invalid GitHub artifacts list response.');
  }
```

Let's avoid this style, it's error-prone.

Reply:
Sure, would you mind explaining/link to some resources as to how/why it is error prone?

Reply:
Also, I'm assuming you didn't write this code then? 😆

https://github.com/electron/website/blob/76c112bea9a8c00d887696fced27bda0d3c76021/.github/workflows/resolve-pr-release-versions.yml#L250

Reply:
Should I change these too?

https://github.com/electron/website/pull/594/files#diff-1dd7a747a324f6f8916a65d87d4c435b9041ed1b0a0d9fa48518423b4d3012edR34-R37

```ts
  const release = prReleaseVersions?.release;
  if (release) allVersions.push(release);

  const backports = prReleaseVersions?.backports;
  if (backports) allVersions.push(...backports);
```